File: leetcode/longest_substring_without_repeating_characters.js
// Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let map = new Map();
    let longest = 0;
    let count = 0;
    let lastIndex = 0;
    for (let i = 0; i < s.length; i++) {
        let char = s.charAt(i);
        let ex = map.get(char);
        if (ex != null && ex >= lastIndex) {
            longest = Math.max(longest,count);
            count = count - ((ex- lastIndex) + 1);
            lastIndex = ex + 1;
        }
        count++
        map.set(char,i);
    }
    if (longest < count) {
        longest = count;
    }
    return longest
};

function main() {
    let res = lengthOfLongestSubstring("abcabcbb")
    console.log(res);
}

main()
let i = 0;
console.log(i++)
console.log(i)

File: leetcode/longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        length = 0
        start = 0
        indexMap = {}
        for i in range(len(s)):
            c = s[i]

            if c in indexMap:

                l = i - start
                if indexMap[c] >= start:
                    start = indexMap[c] + 1

                indexMap[c] = i

                length = max(l, length)

            indexMap[c] = i
        l = i - start + 1
        length = max(l, length)
        return length


if __name__ == '__main__':
    s = "abcaaacbad"
    res = Solution().lengthOfLongestSubstring(s)
    print(res)
